Framework/Kernel/src/TimeSeriesProperty.cpp

```diff
@@ -308,36 +311,63 @@ void TimeSeriesProperty<TYPE>::createFilteredData(const TimeROI &timeROI,
-    // std::lower_bound() and std::upper_bound(), index_stop must already be correct, as long as we treat it
-    // inclusively, but index_start needs to be adjusted if possible.
-    if (index_start > 0)
-      index_start--;
-
-    // handle potential overlaps between successive index intervals, so we don't copy the same datapoints twice
+  // Get all ROI time boundaries. Every other value is start/stop of an ROI "use" region.
+  const std::vector<Types::Core::DateAndTime> &roiTimes = timeROI.getAllTimes();
+  auto itROI = roiTimes.cbegin();
+  const auto itROIEnd = roiTimes.cend();
+
+  auto itValue = m_values.cbegin();
+  const auto itValueEnd = m_values.cend();
+  auto itLastValueUsed = itValue; // last value used up to the moment
+
+  while (itROI != itROIEnd && itValue != itValueEnd) {
+    // Try fast-forwarding the current ROI "use" region towards the current time value. Note, the current value might
+    // be in an ROI "ignore" region together with one or more following values.
+    while (std::distance(itROI, itROIEnd) > 2 && *(std::next(itROI, 2)) <= itValue->time())
+      std::advance(itROI, 2);
+    // Try finding the first value equal or past the beginning of the current ROI "use" region.
+    itValue = std::lower_bound(itValue, itValueEnd, *itROI,
+                               [](const auto &value, const auto &roi_time) { return value.time() < roi_time; });
+    // Calculate a [begin,end) range for the values to use
+    auto itBeginUseValue = itValue;
+    auto itEndUseValue = itValue;
+    // If there are no values past the current ROI "use" region, get the previous value
+    if (itValue == itValueEnd) {
+      itBeginUseValue =
+          std::prev(itValue); // std::prev is safe here, because "m_values is empty" case has already been treated above
+      itEndUseValue = itValueEnd;
+    }
+    // If the value is inside the current ROI "use" region, look for other values in the same ROI "use" region
+    else if (itValue->time() <= *(std::next(itROI))) {
+      // First, try including a value immediately preceding the first value in the ROI "use" region.
+      itBeginUseValue = itValue == m_values.begin() ? itValue : std::prev(itValue);
+      // Now try finding the first value past the end of the current ROI "use" region.
+      while (itValue != itValueEnd && itValue->time() <= *(std::next(itROI)))
+        itValue++;
+      // Include the current value, therefore, advance itEndUseValue, because std::copy works as [begin,end).
+      itEndUseValue = itValue == itValueEnd ? itValue : std::next(itValue);
+    }
+    // If we are at the last ROI "use" region or the value is not past the beginning of the next ROI "use" region, keep
+    // it for the current ROI "use" region.
+    else if (std::distance(itROI, itROIEnd) == 2 ||
+             (std::distance(itROI, itROIEnd) > 2 && itValue->time() < *(std::next(itROI, 2)))) {
```

Comment:
Is `(std::distance(itROI, itROIEnd) > 2` necessary? if the loop works as intended this seems redundant.

Reply:
It is correct. We're advancing `itROI` **only** as long as `*(std::next(itROI, 2)) <= itValue->time()` is true. However, that condition **can't be checked** unless `std::distance(itROI, itROIEnd) > 2`

Reply:
I don't understand your explanation :disappointed:  
If `std::distance(itROI, itROIEnd)` is not `2`, then it must be either `0` or `> 2`. The case `0` you dealt with before, so that leaves `std::distance(itROI, itROIEnd) > 2` being true. Hence, evaluating condition `(std::distance(itROI, itROIEnd) > 2` is a moot point.
Am I missing something here?

Reply:
Could you better explain to me what you mean by "before" in:

> The case 0 you dealt with before, so that leaves std::distance(itROI, itROIEnd) > 2 being true

Are you referring to L336:
`  while (itROI != itROIEnd && itValue != itValueEnd) {
`
If you do, that does not apply here, because it's not the same loop.

Reply:
When loop `while (itROI != itROIEnd && itValue != itValueEnd)` starts executing, we necessarily have `std::distance(itROI, itROIEnd) >= 2`, so at this point we have dealt with condition `std::distance(itROI, itROIEnd) == 0`.
However, I just noticed now the inner `while` loop:
```c++
    while (std::distance(itROI, itROIEnd) > 2 && *(std::next(itROI, 2)) <= itValue->time())
      std::advance(itROI, 2);
```
If this loops brings `itROI` to `itROIEnd`, then we'll have again `std::distance(itROI, itROIEnd) == 0` and in this case we indeed need to check for condition `std::distance(itROI, itROIEnd) > 2`. So, is this inner `while` loop the reason for explicitly introducing condition `std::distance(itROI, itROIEnd) > 2`?

Reply:
Yes, that is the reason.

Reply:
OK, then. This conversation's resolved.
Sorry it took me so long to find out about the inner loop. I guess I wasted your time

Reply:
No, you haven't.  I agree with you, when the inner loop just starts, the distance check is redundant.  However, it's not obvious how to avoid that extra check without introducing even more complexity to the code.

BTW, if you take a look at the next statement, which is a "hidden loop"
```
itValue = std::lower_bound(itValue, itValueEnd, *itROI,
                               [](const auto &value, const auto &roi_time) { return value.time() < roi_time; });
```
there is basically the same inefficiency there, because `std::lower_bound` implementation must exit if `itValue` is `itValueEnd`, but the first check for that is redundant, because we have already checked it in our outer loop!